### PR TITLE
Sdk 172 Cloud Agnostic changes for v2 api cluster requests

### PR DESCRIPTION
**Pull request by tanishgupta1**
* Added cloud as extra variable for qubole configuration setting.

* Created aws , oracle, azure classes to manage cloud config setting, argparser of cluster or other cloud specific functions.

* Added engine class to manage engine specific functions.

* Manage command line arguments accordingly w.r.t cloud. So for aws cloud only aws specific command line arguments will be shown. Similarly for azure and oracle.

### CODE REVIEW

File: bin/qds.py [hunk 1]
```diff
@@ -550,7 +562,14 @@ def main():
...
                      version=options.api_version,
                      poll_interval=options.poll_interval,
-                     skip_ssl_cert_check=options.skip_ssl_cert_check)
+                     skip_ssl_cert_check=options.skip_ssl_cert_check,
+                     cloud=options.cloud)
+
+    # If cloud provider is not set by user at system level or through command line ,
+    # then fetch cloud from api url request
+    if options.cloud is None:
+        options.cloud = Cloud.get_cloud()
```
Comment (sourabh912) [inline bin/qds.py]: We can set default cloud provider to `aws`. I think there is no need to make extra api call just to fetch the cloud provider.

File: qds_sdk/cloud/azure_cloud.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        network_config_group.add_argument("--vnet-name",
+                                          dest="vnet_name",
+                                          help="vnet name for azure", )
+        network_config_group.add_argument("--subnet-name",
+                                          dest="subnet_name",
+                                          help="subnet name for azure")
+        network_config_group.add_argument("--vnet-resource-group-name",
+                                          dest="vnet_resource_group_name",
+                                          help="vnet resource group name for azure")
+        network_config_group.add_argument("--bastion-node-public-dns",
```
Comment (sourabh912) [inline qds_sdk/cloud/azure_cloud.py]: arguments like bastion-node-public-dns, persistent-security-groups, master-elastic-ip are specific to aws I guess. Please verify and remove them from here. Also check arguments passed in oracle cloud.

File: qds_sdk/cloud/cloud.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+from qds_sdk.qubole import Qubole
+
+class Cloud:
+
+    @staticmethod
+    def get_cloud():
```
Comment (sourabh912) [inline qds_sdk/cloud/cloud.py]: Why is `get_cloud` a staticmethod whereas  `get_cloud_object` a classmethod ?

File: qds_sdk/clusterv2.py [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+        Engine().engine_parser(subparser)
+
+
+    @staticmethod
+    def run(args):
+        parser = ClusterCmdLine.parsers(args)
+        arguments = parser.parse_args(args)
+        customer_ssh_key = util._read_file(arguments.customer_ssh_key_file, "customer ssh key file")
+
+        # This will set cluster info and monitoring settings
```
Comment (sourabh912) [inline qds_sdk/clusterv2.py]: Do we need to write parser for cluster actions like list, start, terminate, status, add_node etc ?

File: tests/test_clusterv2.py [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+from mock import Mock
+import tempfile
+sys.path.append(os.path.join(os.path.dirname(__file__), '../bin'))
+import qds
+from qds_sdk.connection import Connection
+from test_base import print_command
+from test_base import QdsCliTestCase
+from qds_sdk.cloud.cloud import Cloud
+
+class TestClusterCreate(QdsCliTestCase):
```
Comment (sourabh912) [inline tests/test_clusterv2.py]: Please add test cases for other cluster actions like `list`, `terminate`, `add_node` etc

File: bin/qds.py [hunk 6]
```diff
@@ -510,6 +518,10 @@ def main():
                          default=False,
                          help="skip verification of server SSL certificate. Insecure: use with caution.")
 
+    optparser.add_option("--cloud", dest="cloud",
+                         default=os.getenv('QDS_CLOUD'),
+                         help="cloud", choices=["aws", "azure", "oracle_bmc"])
```
Comment (ajayaa) [inline bin/qds.py]: @sourabh912 Should this be in capital since these are being set from an environment variable?

**Comment by ajayaa** [comment]
There should be a way to dump the any object and see the json object being constructed.

Right now, there is a single function which takes all the params for constructing a cluster_info, cloud_config, engine_config. Someone using the SDK might not have all the params at one place. To support such use cases, we should support adding an attribute to any object type without having to use the function. IMO we should support both use cases. Someone having all the params at one place and someone constructing it at different places.

File: bin/qds.py [hunk 7]
```diff
@@ -16,6 +16,8 @@
 from qds_sdk.nezha import NezhaCmdLine
 from qds_sdk.user import UserCmdLine
 from qds_sdk.template import TemplateCmdLine
+from qds_sdk.clusterv2 import *
```
Comment (sourabh912) [inline bin/qds.py]: @tanishgupta1 : Please import only required functionality

File: bin/qds.py [hunk 8]
```diff
@@ -238,6 +239,7 @@ def _create_cluster_info(arguments, api_version):
                                       aws_availability_zone=arguments.aws_availability_zone,
                                       vpc_id=arguments.vpc_id,
                                       subnet_id=arguments.subnet_id,
+                                      master_elastic_ip=arguments.master_elastic_ip,
```
Comment (sourabh912) [inline bin/qds.py]: @tanishgupta1 : Is this change required in this PR ? If not, please remove it and raise a separate PR.

File: bin/qds.py [hunk 9]
```diff
@@ -319,6 +324,7 @@ def _create_cluster_info(arguments, api_version):
 
     return cluster_info
 
+
```
Comment (sourabh912) [inline bin/qds.py]: @tanishgupta1 : Please remove these extra lines.

File: bin/qds.py [hunk 10]
```diff
@@ -510,6 +518,10 @@ def main():
                          default=False,
                          help="skip verification of server SSL certificate. Insecure: use with caution.")
 
+    optparser.add_option("--cloud", dest="cloud",
+                         default=os.getenv('QDS_CLOUD'),
```
Comment (sourabh912) [inline bin/qds.py]: Any updates ?

File: bin/qds.py (hunk 1 above)
Comment (sourabh912) [inline bin/qds.py]: Any updates? I think it is better if we set the default cloud provider to `aws`

File: qds_sdk/cloud/aws_cloud.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+                is required for creating the cluster.
+
+            compute_secret_key: The secret access key for customer's aws
+                account. This is required for creating the cluster.
+
+            use_account_compute_creds: Set it to true to use the account’s compute
+                credentials for all clusters of the account.The default value is false
+
+            aws_region: The AWS region in which the cluster is created. The default value is, us-east-1.
+                Valid values are, us-east-1, us-west-1, us-west-2, eu-west-1, sa-east1, ap-southeast-1,
```
Comment (sourabh912) [inline qds_sdk/cloud/aws_cloud.py]: +1

File: qds_sdk/cloud/aws_cloud.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+                              master_elastic_ip=arguments.master_elastic_ip)
+
+    def cloud_config_parser(self, argparser):
+
+        # compute settings parser
+        compute_config = argparser.add_argument_group("compute config settings")
+        compute_creds = compute_config.add_mutually_exclusive_group()
+        compute_creds.add_argument("--enable_account_compute_creds",
+                                   dest="use_account_compute_creds",
+                                action="store_true",
```
Comment (sourabh912) [inline qds_sdk/cloud/aws_cloud.py]: @tanishgupta1 : Please fix indentation.

File: qds_sdk/cloud/cloud.py (hunk 3 above)
Comment (sourabh912) [inline qds_sdk/cloud/cloud.py]: @tanishgupta1 : Any updates ?

File: qds_sdk/clusterv2.py (hunk 4 above)
Comment (sourabh912) [inline qds_sdk/clusterv2.py]: @tanishgupta1 : Any updates ?

File: qds_sdk/util.py [hunk 13]
```diff
@@ -142,3 +142,29 @@ def underscore(word):
...
+            else:
+                new_dict[key] = value
+    return new_dict
+
+
+def _read_file(file_path, file_name):
+    file_content = None
+    if file_path is not None:
+        try:
+            file_content = open(file_path).read()
```
Comment (sourabh912) [inline qds_sdk/util.py]: @tanishgupta1 : Open file using `with` statement. It takes care of properly closing file descriptors.

Comment (sourabh912) [inline qds_sdk/util.py]: Also I believe, this function should throw an exception and its caller should handle that exception.

File: qds_sdk/cloud/aws_cloud.py [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+            self.network_config['persistent_security_groups'] = persistent_security_groups
+            self.network_config['master_elastic_ip'] = master_elastic_ip
+            self.network_config['vpc_id'] = vpc_id
+            self.network_config['subnet_id'] = subnet_id
+
+        set_compute_config()
+        set_location()
+        set_network_config()
+
+    def set_cloud_config_settings(self, arguments):
```
Comment (sourabh912) [inline qds_sdk/cloud/aws_cloud.py]: Rename it to `set_cloud_config_from_arguments`

File: qds_sdk/cloud/aws_cloud.py [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+                              aws_region=arguments.aws_region,
+                              aws_availability_zone=arguments.aws_availability_zone,
+                              role_instance_profile=arguments.role_instance_profile,
+                              vpc_id=arguments.vpc_id,
+                              subnet_id=arguments.subnet_id,
+                              persistent_security_groups=arguments.persistent_security_groups,
+                              bastion_node_public_dns=arguments.bastion_node_public_dns,
+                              master_elastic_ip=arguments.master_elastic_ip)
+
+    def cloud_config_parser(self, argparser):
```
Comment (sourabh912) [inline qds_sdk/cloud/aws_cloud.py]: Rename it to maybe `create_parser`.

File: qds_sdk/cloud/cloud.py [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+        conn = Qubole.agent(version="v1.2")
+        url_path = "about"
+        return conn.get(url_path).get("provider")
+
+    @staticmethod
+    def get_cloud_object():
```
Comment (sourabh912) [inline qds_sdk/cloud/cloud.py]: Rename it to : `get_cloud`.

File: qds_sdk/clusterv2.py [hunk 17]
```diff
@@ -0,0 +1,550 @@
...
+import json
+
+def str2bool(v):
+    return v.lower() in ("yes", "true", "t", "1")
+
+class ClusterCmdLine:
+
+    @staticmethod
+    def parsers(args):
+        action = args[0]
```
Comment (sourabh912) [inline qds_sdk/clusterv2.py]: Pass only action and then create an appropriate parser based on action type.

File: tests/test_clusterv2.py (hunk 5 above)
Comment (tanishgupta1) [inline tests/test_clusterv2.py]: Update: Function other than update, create and clone will fall under v1.3 api version even if the version given by user is v2.

File: bin/qds.py (hunk 9 above)
Comment (tanishgupta1) [inline bin/qds.py]: removed

File: qds_sdk/cloud/aws_cloud.py [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+                Typically used to provide access to external hosts
+
+            bastion_node_public_dns: Specify the Bastion host public DNS name if private subnet is provided.
+                Do not specify this value for a public subnet.
+
+            master_elastic_ip: It is the Elastic IP address for attaching to the cluster master
+
+        '''
+
+        def set_compute_config():
```
Comment (tanishgupta1) [inline qds_sdk/cloud/aws_cloud.py]: Do we need to create separate functions for this instead of calling it internally through set_cloud_config?

**Comment by harshshah87** [comment]
Looks like this has been merged. Please re-open if required.
